feat(learn): Add tests to Personal Library

SaintPeter opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

### Changes
* Adds all of the tests for all user stories, including my improved user stories (See #39630 and PR https://github.com/freeCodeCamp/boilerplate-project-library/pull/5)
~~* Removes InfoSec tests as per #39435~~

### Testing
I have tested this against my personal project. It will fail against the current boilerplate.

SaintPeter committed:
Resolve Merge Conflict

SaintPeter commented:
I should have resolved the merge conflict.

naomi-lgbt reviewed (dismissed):
This LGTM!

I tested locally against @SaintPeter's project, as the current example does fail. I am, however, in favour of adding the additional clarity to these test instructions. 

As such, prior to merging this I recommend ensuring an update on the example project.

SaintPeter commented:
If you want to test against my solution:
https://personal-library.pearlygatesoftware.com/

raisedadead commented:
This is blocked by #39024

RandellDawson commented:
Fixed merge conflict, and PR #39024 has been merged, so removing `status: blocked` label.

SaintPeter committed:
Add Instructions, Reformat Description

SaintPeter commented:
Fixed the merge conflict, added the instructions from the README file, as well as removing the user stories as per https://github.com/freeCodeCamp/freeCodeCamp/issues/39721.

Sky020 commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -7,14 +7,22 @@ forumTopicId: 301571
 <section id='description'>
-Build a full stack JavaScript app that is functionally similar to this: <a href="https://personal-library.freecodecamp.rocks/" target="_blank">https://personal-library.freecodecamp.rocks/</a>.
-Working on this project will involve you writing your code on Repl.it on our starter project. After completing this project you can copy your public Repl.it URL (to the homepage of your app) into this screen to test it! Optionally you may choose to write your project on another platform but must be publicly visible for our testing.
+Build a full stack JavaScript app that is functionally similar to this: <a href="https://personal-library--freecodecamp.repl.co/" target="_blank">https://personal-library--freecodecamp.repl.co/</a>.
+Working on this project will involve you writing your code on Repl.it on our starter project. After completing this project you can copy your public Repl.it URL (to the homepage of your app) into this screen to test it! Optionally you may choose to write your project on another platform but must be publicly visible for our testing.  
+<br><br>
 Start this project on Repl.it using <a href="https://repl.it/github/freeCodeCamp/boilerplate-project-library/">this link</a> or clone <a href='https://github.com/freeCodeCamp/boilerplate-project-library/'>this repository</a> on GitHub! If you use Repl.it, remember to save the link to your project somewhere safe!
```
```suggestion

Build a full stack JavaScript app that is functionally similar to this: <a href='https://personal-library.freecodecamp.rocks/' target='_blank'>https://personal-library.freecodecamp.rocks/</a>. Working on this project will involve you writing your code using one of the following methods:

- Clone <a href='https://github.com/freeCodeCamp/boilerplate-personal-library'>this GitHub repo</a> and complete your project locally.
- Use <a href='https://repl.it/github/freeCodeCamp/boilerplate-personal-library' target='_blank'>our repl.it starter project</a> to complete your project.
- Use a site builder of your choice to complete the project. Be sure to incorporate all the files from our GitHub repo.

When you are done, submit a URL to a publicly visible working demo of your project in the `Solution Link` field and, optionally, submit a link to your projects source code in the `GitHub Link` field.

```

moT01 replied:
Hey, sorry about that - I made one more comment regarding these to try and clarify the wording a little more. We may want to change it up a tiny bit yet.

ShaunSHamilton reviewed (dismissed):
The description for the projects has changed: 39721

SaintPeter committed:
Update curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md

Co-authored-by: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

moT01 commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -7,14 +7,27 @@ forumTopicId: 301571
+- Use <a href='https://repl.it/github/freeCodeCamp/boilerplate-personal-library' target='_blank'>our repl.it starter project</a> to complete your project.
+- Use a site builder of your choice to complete the project. Be sure to incorporate all the files from our GitHub repo.
+
+When you are done, submit a URL to a publicly visible working demo of your project in the `Solution Link` field and, optionally, submit a link to your projects source code in the `GitHub Link` field.
```
```suggestion
When you are done, make sure a working demo of your project is hosted somewhere public. Then submit the URL to it in the `Solution Link` field. Optionally, also submit a link to your projects source code in the `GitHub Link` field.
```

moT01 replied:
Last one

SaintPeter committed:
Update curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md

Co-authored-by: Tom <20648924+moT01@users.noreply.github.com>

moT01 commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -7,14 +7,27 @@ forumTopicId: 301571
+- Use <a href='https://repl.it/github/freeCodeCamp/boilerplate-personal-library' target='_blank'>our repl.it starter project</a> to complete your project.
+- Use a site builder of your choice to complete the project. Be sure to incorporate all the files from our GitHub repo.
+
+When you are done, make sure a working demo of your project is hosted somewhere public. Then submit the URL to it in the `Solution Link` field. Optionally, also submit a link to your projects source code in the `GitHub Link` field.
```
```suggestion
When you are done, make sure a working demo of your project is hosted somewhere public. Then submit the URL to it in the `Solution Link` field. Optionally, also submit a link to your project's source code in the `GitHub Link` field.
```

moT01 committed:
Update curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md

Sky020 commented:
Are the user stories going to be added to /learn in this PR?

SaintPeter commented:
@Sky020 This PR adds the tests to the existing user stories on `/learn`.

Sky020 commented:
@SaintPeter Right. I was looking at the _most recent_ in the commit history.

Still, I was expecting to see the same _user stories_ which were on the `index.html` page. I am concerned many campers will not know where to begin, if all they see is the tests.

SaintPeter commented:
@Sky020 While the most recent commit alters the instructions, the prior commits add tests to the existing user stories. As I understand it, the new standard is for the only instance of the user stories will be on the `/learn` page... Which they are.

It might make sense to include a specific instruction to the instructions section pointing them to the User Stories / Tests below the project submission buttons. Maybe a re-layout of the project page is needed, in light of this new standard?

Sky020 commented:
@SaintPeter I was thinking of [this comment](https://github.com/freeCodeCamp/freeCodeCamp/issues/39721#issuecomment-700703006):
> I think that, for now, we would be fine doing that and once we get everything moved over to learn, perhaps we can look into moving some user stories to the instructions area.

So, ignore my comments, for now. We should discuss it more officially, before doing anything. Thanks, for bearing with my scatter-brain.

SaintPeter committed:
Change links to Markdown style

SaintPeter commented:
Figured out what was blocking the linter in the CI and updated the HTML style links to Markdown style links.

nhcarrigan commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -27,23 +40,134 @@ tests:
+        throw new Error(err.responseText || err.message);
+      }
+    }"
+  - text: I can <b>get</b> /api/books/{_id} to retrieve a single object of a book containing <code>title</code>, <code>_id</code>, & an array of <code>comments</code> (empty array if no comments present).
```
```suggestion
  - text: I can <b>get</b> /api/books/{_id} to retrieve a single object of a book containing <code>title</code>, <code>_id</code>, & an array of <code>comments</code> (empty array if no comments present). If no book is found, return '<code>no book exists</code>'.
```

nhcarrigan commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -27,23 +40,134 @@ tests:
+        throw new Error(err.responseText || err.message);
+      }
+    }"
+  - text: I can <b>post</b> a <code>comment</code> to /api/books/{_id} to add a comment to a book and returned will be the books object similar to <b>get</b> /api/books/{_id}. If <code>comment</code> is not included in the request, return '<code>missing required field comment</code>'
```
```suggestion
  - text: I can <b>post</b> a <code>comment</code> to /api/books/{_id} to add a comment to a book and returned will be the books object similar to <b>get</b> /api/books/{_id}. If <code>comment</code> is not included in the request, return '<code>missing required field comment</code>'. If no book is found, return '<code>no book exists</code>`.
```

nhcarrigan commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -27,23 +40,134 @@ tests:
+        throw new Error(err.responseText || err.message);
+      }
+    }"
+  - text: I can <b>delete</b> /api/books/{_id} to delete a book from the collection. Returned will be '<code>delete successful</code>' if successful.
```
```suggestion
  - text: I can <b>delete</b> /api/books/{_id} to delete a book from the collection. Returned will be '<code>delete successful</code>' if successful. If no book is found, return '<code>no book exists</code>'.
```

nhcarrigan commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -27,23 +40,134 @@ tests:
+         let commentErr = await $.post(url + '/' + bookId);
+         assert.isString(commentErr);
+         assert.equal(commentErr, 'missing required field comment');
+         let failingComment = await $.post(url + '/f665eb46e296f6b9b6a504d', { 'comment': 'Never Seen Comment' });
```
```suggestion
         let failingComment = await $.post(url + '/5f665eb46e296f6b9b6a504d', { 'comment': 'Never Seen Comment' });
```
Missed the `5` here so doesn't cast to _id properly.

naomi-lgbt reviewed (dismissed):
Some things I ran in to while working on an example project for this.

SaintPeter committed:
Apply suggestions from code review

Co-authored-by: Nicholas Carrigan (he/him) <nhcarrigan@gmail.com>

RandellDawson commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -27,23 +40,134 @@ tests:
-  - text: All 6 functional tests required are complete and passing.
-    testString: ''
-
+  - text: I can <b>post</b> a <code>title</code> to /api/books to add a book and returned will be the object with the <code>title</code> and a unique <code>_id</code>.  If <code>title</code> is not included in the request, return '<code>missing required field title</code>'
```
```suggestion
  - text: You can send a <b>POST</b> request to <code>/api/books</code> with <code>title</code> as part of the form data to add a book.  The returned response will be an object with the <code>title</code> and a unique <code>_id</code> as keys.  If <code>title</code> is not included in the request, the returned response should be the string <code>missing required field title</code>.
```

RandellDawson commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -27,23 +40,134 @@ tests:
+        throw new Error(err.responseText || err.message);
+      }
+    }"
+  - text: I can <b>get</b> /api/books to retrieve an array of all books containing <code>title</code>, <code>_id</code>, & <code>commentcount</code>.
```
```suggestion
  - text: You can send a <b>GET</b> request to <code>/api/books</code> and receive a JSON response representing all the books. 
 The JSON response will be an array of objects with each object (book) containing <code>title</code>, <code>_id</code>, and <code>commentcount</code> properties.
```

RandellDawson commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -27,23 +40,134 @@ tests:
+        throw new Error(err.responseText || err.message);
+      }
+    }"
+  - text: I can <b>get</b> /api/books/{_id} to retrieve a single object of a book containing <code>title</code>, <code>_id</code>, & an array of <code>comments</code> (empty array if no comments present). If no book is found, return '<code>no book exists</code>'.
```
```suggestion
  - text: You can send a <b>GET</b> request to <code>/api/books/{_id}</code> to retrieve a single object of a book containing the properties <code>title</code>, <code>_id</code>, and a <code>comments</code> array (empty array if no comments present). If no book is found, return the string <code>no book exists</code>.
```

RandellDawson commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -27,23 +40,134 @@ tests:
+        throw new Error(err.responseText || err.message);
+      }
+    }"
+  - text: I can <b>post</b> a <code>comment</code> to /api/books/{_id} to add a comment to a book and returned will be the books object similar to <b>get</b> /api/books/{_id}. If <code>comment</code> is not included in the request, return '<code>missing required field comment</code>'. If no book is found, return '<code>no book exists</code>`.
```
```suggestion
  - text: You can send a <b>POST</b> request containing <code>comment</code> as the form body data to <code>/api/books/{_id}</code> to add a comment to a book. The returned response will be the books object similar to <b>GET</b> <code>/api/books/{_id}</code> request in an earlier test. If <code>comment</code> is not included in the request, return the string <code>missing required field comment</code>`. If no book is found, return the string <code>no book exists</code>.
```

RandellDawson commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -27,23 +40,134 @@ tests:
+        throw new Error(err.responseText || err.message);
+      }
+    }"
+  - text: I can <b>delete</b> /api/books/{_id} to delete a book from the collection. Returned will be '<code>delete successful</code>' if successful. If no book is found, return '<code>no book exists</code>'.
```
```suggestion
  - text: You can send a <b>DELETE</b> request to <code>/api/books/{_id}</code> to delete a book from the collection. The returned response will be the string <code>delete successful</code> if successful. If no book is found, return the string <code>no book exists</code>.
```

RandellDawson commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -27,23 +40,134 @@ tests:
+        throw new Error(err.responseText || err.message);
+      }
+    }"
+  - text: I can send a <b>delete</b> request to /api/books to delete all books in the database. Returned will be '<code>'complete delete successful</code>' if successful.
```
```suggestion
  - text: You can send a <b>DELETE</b> request to <code>/api/books</code> to delete all books in the database. The returned response will be the string <code>'complete delete successful</code> if successful.
```

RandellDawson reviewed (changes requested):
@SaintPeter @nhcarrigan I tested the changes against the changes in the example project PR (https://github.com/freeCodeCamp/demo-projects/pull/39) and everything looks fine.  I do have a few suggested changes. If you have a different way of wording what I have suggested, please feel free to let me know and I can update my suggestion on GitHub.

In addition to the suggested changes below, can you update the first test's `text` to the following to bring this project in line with the suggestions of issue https://github.com/freeCodeCamp/freeCodeCamp/issues/39721:
> You should provide your own project, not the example URL.

Commit:
Apply suggestions from code review

Co-authored-by: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>

Commit:
Verbiage per #39721

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

RandellDawson commented:
I am good with these changes now.  @moT01 @Sky020 @nhcarrigan If you are good, then we can get this one merged.

nhcarrigan commented:
I built an example project against this change and everything seemed good to me. I'm good with this.

RandellDawson reviewed (dismissed):
@SaintPeter For the most part I am satisfied with the current changes.  Will still need to create a separate PR for this project and the boilerplate after merging both this PR and the [existing boilerplate PR](https://github.com/freeCodeCamp/boilerplate-project-library/pull/5) to add any further instructions that are present in the boilerplate comments and then remove those comments from the boilerplate to complete the requirements specified in issue https://github.com/freeCodeCamp/freeCodeCamp/issues/39721 .  I recommend looking at what was done in the Mocha/Chai challenges PR (#39976) by using `#` to reference the specific test the user should modify in the boilerplate.

For now, I am just have one suggested change for this PR.

moT01 reviewed (commented):
Overall, these look real good @SaintPeter & @nhcarrigan 🎉 I am just going to leave my findings for all three PR's related to this project here.

I was going to mention what Randell brought up as well.

The boilerplate has some files in it that could be deleted - `unit-tests.js`, `.bashrc`, `.profile`, `.hyperdev-assets`. I'm quite sure these aren't needed - they don't have to be deleted before these can get merged - I will leave it up to you two if you want to do that. `unit-tests.js` also exists on the demo-project.

The demo project hangs for me when making any type of request to a bad id - e.g. https://personal-library.nicholascarriga.repl.co/api/books/bad_id - It does seem to work, however, if the "bad_id" is an id of a book that was deleted. So if I create a book, then delete it and use that id, it will give the correct `no book exists` response. I don't think this is a huge deal, but it might be nice to have the demo give out those responses. It also means that the tests are only checking ids that formerly existed or something? not sure. Thought I would point it out if you two want to take a look at all - perhaps you're not seeing the same thing. 

I would be fine merging all three of the PR's as-is if we want 🎉

nhcarrigan commented:
> The demo project hangs for me when making any type of request to a bad id - e.g. https://personal-library.nicholascarriga.repl.co/api/books/bad_id - It does seem to work, however, if the "bad_id" is an id of a book that was deleted. So if I create a book, then delete it and use that id, it will give the correct no book exists response. I don't think this is a huge deal, but it might be nice to have the demo give out those responses. It also means that the tests are only checking ids that formerly existed or something? not sure. Thought I would point it out if you two want to take a look at all - perhaps you're not seeing the same thing.

This is likely a MongoDB "failed to cast to ObjectId" error - if the string isn't something that looks like an ObjectID, it gets mad. 
I can poke at this later the afternoon and update the demo project accordingly. 🙂

nhcarrigan commented:
@moT01 I've added a check for valid ObjectID to the example project, which should resolve the error you encountered.

moT01 commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -7,43 +7,167 @@ forumTopicId: 301571
+   Example: `DB=mongodb://admin:pass@1234.mlab.com:1234/fccpersonallib`
+2. In your `.env` file set `NODE_ENV` to `test`, without quotes
+3. You need to create all routes within `routes/api.js`
+4. You will create all unit tests in `tests/1_unit-tests.js`
```
may want to change the 5 below this to 4 as well? - or you can also just turn all of them into 1 for an ordered list.

SaintPeter committed:
Update curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md

Co-authored-by: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>

SaintPeter commented:
I've made the requested changes.

nhcarrigan commented on curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md:
```diff
@@ -7,43 +7,166 @@ forumTopicId: 301571
 ## Instructions
 <section id='instructions'>
 
+1. Add your MongoDB connection string to `.env` without quotes as `DB`  
+   Example: `DB=mongodb://admin:pass@1234.mlab.com:1234/fccpersonallib`
+2. In your `.env` file set `NODE_ENV` to `test`, without quotes
+3. You need to create all routes within `routes/api.js`
+5. You will create all functional tests in `tests/2_functional-tests.js`
```
```suggestion
1. Add your MongoDB connection string to `.env` without quotes as `DB`  
   Example: `DB=mongodb://admin:pass@1234.mlab.com:1234/fccpersonallib`
2. In your `.env` file set `NODE_ENV` to `test`, without quotes
3. You need to create all routes within `routes/api.js`
4. You will create all functional tests in `tests/2_functional-tests.js`
```

While markdown renders this correctly, it appears the linter doesn't like it.

SaintPeter committed:
Update curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/personal-library.md

Co-authored-by: Nicholas Carrigan (he/him) <nhcarrigan@gmail.com>

moT01 reviewed (approved):
LGTM 🎉 Thanks for fixing this up @SaintPeter @nhcarrigan 🎉
